File: server/utils/redisClient.js
/**
 * redisClient.js — ioredis-based Redis utility for the Node.js backend.
 *
 * Provides:
 *   getRedis()              → ioredis instance (lazy singleton, async)
 *   setPendingOut(personId, data)   → create / overwrite pending OUT entry
 *   getPendingOut(personId)         → read a single pending OUT entry
 *   getAllPendingOuts()              → scan all pending_out:* keys
 *   resetPendingOut(personId)       → reset TTL + checkCount for re-detection
 *   deletePendingOut(personId)      → remove a pending OUT key
 *
 * Falls back gracefully if Redis is unavailable — the server never crashes.
 */

import Redis from "ioredis";

// ── Constants ───────────────────────────────────────────────────────────
const PENDING_OUT_PREFIX = "pending_out:";
const PENDING_OUT_TTL = 60; // Increased to 60s to ensure all 3 cross-checks can complete

// ── Singleton connection ────────────────────────────────────────────────
let _redis = null;
let _connectAttempted = false;

/**
 * Return a lazy-initialised ioredis connection (async).
 * Returns null if Redis is unavailable — all helpers gracefully no-op.
 */
export async function getRedis() {
    if (_redis) return _redis;
    if (_connectAttempted) return null; // Already failed once, don't retry every call

    _connectAttempted = true;
    const redisUrl = process.env.REDIS_URL || "redis://localhost:6379";

    try {
        const client = new Redis(redisUrl, {
            maxRetriesPerRequest: 3,
            lazyConnect: true,          // Don't auto-connect in constructor
            connectTimeout: 5000,
            retryStrategy(times) {
                if (times > 5) return null; // Stop retrying after 5 attempts
                return Math.min(times * 500, 3000);
            },
        });

        // Catch error events so they don't crash the process
        client.on("error", (err) => {
            console.error(`❌ [Redis] Connection error: ${err.message}`);
        });

        await client.connect();
        console.log(`✅ [Redis] Connected: ${redisUrl}`);
        _redis = client;
        return _redis;
    } catch (err) {
        console.warn(`⚠️ [Redis] Unavailable (${err.message}) — running without Redis`);
        _redis = null;
        return null;
    }
}

// ── Pending OUT helpers ─────────────────────────────────────────────────

/**
 * Create or overwrite a pending OUT entry in Redis.
 * @param {string} personId - Employee / person MongoDB ID
 * @param {object} data - { personId, detectedAt, expiresAt, checkCount, activeCameras }
 * @returns {Promise<boolean>}
 */
export async function setPendingOut(personId, data) {
    const redis = await getRedis();
    if (!redis) return false;

    try {
        const key = `${PENDING_OUT_PREFIX}${personId}`;
        await redis.setex(key, PENDING_OUT_TTL, JSON.stringify(data));
        return true;
    } catch (err) {
        console.error(`❌ [Redis] setPendingOut failed for ${personId}: ${err.message}`);
        return false;
    }
}

/**
 * Get a single pending OUT entry.
 * @param {string} personId
 * @returns {Promise<object|null>}
 */
export async function getPendingOut(personId) {
    const redis = await getRedis();
    if (!redis) return null;

    try {
        const key = `${PENDING_OUT_PREFIX}${personId}`;
        const raw = await redis.get(key);
        return raw ? JSON.parse(raw) : null;
    } catch (err) {
        console.error(`❌ [Redis] getPendingOut failed for ${personId}: ${err.message}`);
        return null;
    }
}

/**
 * Scan for ALL pending_out:* keys and return parsed values.
 * Uses SCAN to avoid blocking Redis on large datasets.
 * @returns {Promise<object[]>} - Array of parsed pending OUT entries
 */
export async function getAllPendingOuts() {
    const redis = await getRedis();
    if (!redis) return [];

    try {
        const results = [];
        let cursor = "0";

        do {
            const [nextCursor, keys] = await redis.scan(cursor, "MATCH", `${PENDING_OUT_PREFIX}*`, "COUNT", 100);
            cursor = nextCursor;

            if (keys.length > 0) {
                // Pipeline GET for all found keys
                const pipeline = redis.pipeline();
                keys.forEach((key) => pipeline.get(key));
                const values = await pipeline.exec();

                for (const [err, raw] of values) {
                    if (!err && raw) {
                        try {
                            results.push(JSON.parse(raw));
                        } catch {
                            // skip malformed entries
                        }
                    }
                }
            }
        } while (cursor !== "0");

        return results;
    } catch (err) {
        console.error(`❌ [Redis] getAllPendingOuts failed: ${err.message}`);
        return [];
    }
}

/**
 * Reset the TTL and checkCount for a pending OUT entry.
 * Used when the person is re-detected on the OUT camera.
 * @param {string} personId
 * @returns {Promise<boolean>}
 */
export async function resetPendingOut(personId) {
    const redis = await getRedis();
    if (!redis) return false;

    try {
        const key = `${PENDING_OUT_PREFIX}${personId}`;
        const raw = await redis.get(key);
        if (!raw) return false;

        const data = JSON.parse(raw);
        data.checkCount = 0;
        data.expiresAt = new Date(Date.now() + 30 * 1000).toISOString();

        await redis.setex(key, PENDING_OUT_TTL, JSON.stringify(data));
        return true;
    } catch (err) {
        console.error(`❌ [Redis] resetPendingOut failed for ${personId}: ${err.message}`);
        return false;
    }
}

/**
 * Delete a pending OUT key (used on cancel or confirm).
 * @param {string} personId
 * @returns {Promise<boolean>}
 */
export async function deletePendingOut(personId) {
    const redis = await getRedis();
    if (!redis) return false;

    try {
        const key = `${PENDING_OUT_PREFIX}${personId}`;
        await redis.del(key);
        return true;
    } catch (err) {
        console.error(`❌ [Redis] deletePendingOut failed for ${personId}: ${err.message}`);
        return false;
    }
}

File: face-engine/redis_client.py
"""
Redis client — shared connection for the face-engine.

Provides:
    get_redis()  →  redis.Redis (lazy singleton)
    cache_get(key) → dict | None
    cache_set(key, value, ttl)
    cache_delete(key)

Falls back gracefully if Redis is unavailable so the service
never crashes due to a cache miss.
"""

import json
import logging
import os
import redis

logger = logging.getLogger(__name__)

_redis_client: redis.Redis | None = None


def get_redis() -> redis.Redis | None:
    """Return a lazy-initialised Redis connection (or None if unavailable)."""
    global _redis_client
    if _redis_client is not None:
        return _redis_client

    redis_url = os.getenv("REDIS_URL", "redis://localhost:6379")
    try:
        _redis_client = redis.from_url(
            redis_url,
            decode_responses=True,
            socket_connect_timeout=3,
            socket_timeout=2,
            retry_on_timeout=True,
        )
        _redis_client.ping()
        logger.info("✅ Redis connected: %s", redis_url)
    except Exception as e:
        logger.warning("⚠️  Redis unavailable (%s) — running without cache", e)
        _redis_client = None

    return _redis_client


# ── Convenience helpers ──────────────────────────────────────────────────


def cache_get(key: str) -> dict | None:
    """Get a JSON-serialised value from Redis. Returns None on miss or error."""
    r = get_redis()
    if r is None:
        return None
    try:
        raw = r.get(key)
        if raw is None:
            return None
        return json.loads(raw)
    except Exception as e:
        logger.debug("Redis GET %s failed: %s", key, e)
        return None


def cache_set(key: str, value, ttl: int = 30) -> bool:
    """Set a JSON-serialised value in Redis with TTL (seconds)."""
    r = get_redis()
    if r is None:
        return False
    try:
        r.setex(key, ttl, json.dumps(value))
        return True
    except Exception as e:
        logger.debug("Redis SET %s failed: %s", key, e)
        return False


def cache_delete(key: str) -> bool:
    """Delete a key from Redis."""
    r = get_redis()
    if r is None:
        return False
    try:
        r.delete(key)
        return True
    except Exception as e:
        logger.debug("Redis DEL %s failed: %s", key, e)
        return False


def cache_delete_pattern(pattern: str) -> int:
    """Delete all keys matching a glob pattern. Returns count deleted."""
    r = get_redis()
    if r is None:
        return 0
    try:
        keys = list(r.scan_iter(match=pattern, count=200))
        if keys:
            return r.delete(*keys)
        return 0
    except Exception as e:
        logger.debug("Redis DEL pattern %s failed: %s", pattern, e)
        return 0
